pkg/api/query/v1.go

```diff
@@ -1049,22 +1071,18 @@ func (qapi *QueryAPI) labelValues(r *http.Request) (interface{}, []error, *api.A
 		return nil, nil, apiErr, func() {}
 	}
 
-	var matcherSets [][]*labels.Matcher
-	for _, s := range r.Form[MatcherParam] {
-		matchers, err := parser.ParseMetricSelector(s)
-		if err != nil {
-			return nil, nil, &api.ApiError{Typ: api.ErrorBadData, Err: err}, func() {}
-		}
-		matcherSets = append(matcherSets, matchers)
-	}
-
 	tenant, err := tenancy.GetTenantFromHTTP(r, qapi.tenantHeader, qapi.defaultTenant, qapi.tenantCertField)
 	if err != nil {
 		apiErr = &api.ApiError{Typ: api.ErrorBadData, Err: err}
 		return nil, nil, apiErr, func() {}
 	}
 	ctx = context.WithValue(ctx, tenancy.TenantKey, tenant)
 
+	matcherSets, apiErr := qapi.getLabelMatchers(r.Form[MatcherParam], tenant)
+	if apiErr != nil {
+		return nil, nil, apiErr, func() {}
+	}
```

Comment:
Why we don't use `EnforceQueryTenancy` here like in the `query` function? Seems like we have logic to inject the labels in two different places. I would much rather prefer that we have it only in one.

Reply:
Because the APIs are different. The query APIs takes a single PromQL string (which is parsed in prom-label-proxy first, before adjust the labels), while the non-query APIs can have multiple label matchers provided with the `match[]` argument.

Reply:
A query is a label matcher. 😄 

`my_metric{}` is `{__name__="my_metric"`}.

Reply:
They can be parsed in the same way, with `parser.ParseExpr`.

Reply:
I've done a refactor of this part, and now we're doing this:

For query APIs we use prom-label-proxies `EnforceNode` which parses the PromQL string, and add labels
For non-query APIs we parse the matchers manually (which we need regardless of tenancy enforcement), and then use the `EnforceMatchers` from prom-label-proxy to enforce labels.

Hopefully that is an OK solution for you.